### PR TITLE
core,udev,networkd: add ConditionKernelVersion=

### PR DESCRIPTION
**Pull request by poettering**
This adds a simple condition/assert/match to the service manager, to
udev's .link handling and to networkd, for matching the kernel version
string.

In this version we only do fnmatch() based globbing, but we might want
to extend that to version comparisons later on, if we like, by slightly
extending the syntax with ">=", "<=", ">", "<" and "==" expressions.

### CODE REVIEW

**Comment by mbiebl** [comment]
Do we have a use case for this? Testing for specific kernel versions is almost always a bad idea.

**Comment by poettering** [comment]
@mbiebl yeah, we had discussions internally at RH about being able to ship .link files and such inside of kernel packages and make sure they only apply to the kernel they are shipped with.

**Comment by fsateler** [comment]
Hmm. This feature sounds like a very distro-specific thing. After all, kernel versions only mean something within a specific distro, and only if you use distro kernels.

**Comment by poettering** [comment]
> Hmm. This feature sounds like a very distro-specific thing. After all, kernel versions only mean something within a specific distro, and only if you use distro kernels.

Well, fedora/rhel style kernels report something like "4.14.0-1.fc28.x86_64", i.e. an identifier for the distro and such. using that to bind link files to kernels, and shipping them along kernel packages makes a lot of sense...

**Comment by mbiebl** [comment]
@poettering we should at least make it crystal clear then, where it makes sort of sense to use ConditionKernelVersion= to give users clear guidance when it's ok to use or not (still feel like this is a hack tbh)

**Comment by mbiebl** [comment]
The reason being, that it makes unit files unportable across distros.

File: man/systemd.link.xml [hunk 1]
```diff
@@ -172,6 +172,17 @@
           for details.</para>
         </listitem>
       </varlistentry>
+        <varlistentry>
+          <term><varname>KernelVersion=</varname></term>
+          <listitem>
+            <para>Checks whether the kernel version (as reported by <command>uname -r</command>) matches a certain
+            expression (or if prefixed with the exclamation mark does not match it). See
+            <literal>ConditionKernelCommandLine=</literal> in
```
Comment (keszybz) [inline man/systemd.link.xml]: ConditionKernelVersion=

File: man/systemd.netdev.xml [hunk 2]
```diff
@@ -232,6 +232,16 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>KernelVersion=</varname></term>
+        <listitem>
+          <para>Checks whether the kernel version (as reported by <command>uname -r</command>) matches a certain
+          expression (or if prefixed with the exclamation mark does not match it). See
+          <literal>ConditionKernelCommandLine=</literal> in
```
Comment (keszybz) [inline man/systemd.netdev.xml]: ConditionKernelVersion=

File: man/systemd.network.xml [hunk 3]
```diff
@@ -193,6 +193,17 @@
             </para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>KernelVersion=</varname></term>
+          <listitem>
+            <para>Checks whether the kernel version (as reported by <command>uname -r</command>) matches a certain
+            expression (or if prefixed with the exclamation mark does not match it). See
+            <literal>ConditionKernelCommandLine=</literal> in
```
Comment (keszybz) [inline man/systemd.network.xml]: ConditionKernelVersion=

**Comment by poettering** [comment]
@mbiebl note btw, that we already key kmod loading off the same -r field (so that you can have multiple versions of kernel modules around, for each kernel), and it is also used to match up initrds with their kernels. Hence I think keying off other stuff, such as .link files or even units the same way is OK. And to my knowledge Debian does all this the same way as Fedora/RHEL in this regard, no?

**Comment by poettering** [comment]
> ConditionKernelVersion=*.fc28.* # current proposal ConditionKernelVersion=*.fc28.*, *.fc27.* # multiple patterns ConditionKernelVersion= >=1.2.3, <1.2.8 # version comparisons

Yeah, something like that. Though we'd have to figure out whether multiple listed patterns means AND or OR... For the pattern matching OR sounds more appropriate while for the relative comparisons AND makes more sense...

Given that Conditions are ANDed anyway by default if multiple are specified, maybe it should mean OR

**Comment by keszybz** [comment]
Yeah, OR sounds better. This means we should add some special symbol for AND. Maybe just "and"? `ConditionKernelVersion= >=1.2.3 and < 1.3` ?

**Comment by poettering** [comment]
i#d not add any special concept for AND, people should just use multiple conditions:

```
ConditionKernelVersion= >=1.2.3 
ConditionKernelVersion= < 1.3
```

**Comment by keszybz** [comment]
You're right. This will work. Can you fix the docs references so we can merge this?

**Comment by mbiebl** [comment]
> @mbiebl note btw, that we already key kmod loading off the same -r field (so that you can have multiple versions of kernel modules around, for each kernel), and it is also used to match up initrds with their kernels. Hence I think keying off other stuff, such as .link files or even units the same way is OK. And to my knowledge Debian does all this the same way as Fedora/RHEL in this regard, no?

Can you be more specific, so I can check?

**Comment by mbiebl** [comment]
> I think this can go in. Indeed it is mostly useful for local or distro-specific hacks, but not every feature is about being portable, some features are about local tweaks, and that is OK.

Shouldn't we at least warn about this in the docs? I.e. when it's ok to use or not.
I would hate to see if suddenly .service files popped up which make use of that.

**Comment by poettering** [comment]
> Can you be more specific, so I can check?

kmod will look for kernel modules here:

```
# ls /lib/modules/`uname -r`
```

**Comment by mbiebl** [comment]
@poettering not sure, but how is that relevant to systemd and matching to specific kernel versions?
That kmod looks into kernel specific versions for modules makes sense...

**Comment by keszybz** [comment]
I force-pushed a version which fixes the references in the man page, and adds a comment which addresses unportability of this setting. @mbiebl PTAL.

**Comment by poettering** [comment]
> @poettering not sure, but how is that relevant to systemd and matching to specific kernel versions?
> That kmod looks into kernel specific versions for modules makes sense...

Well, it's the same thing: using "uname -r" is the right way to bind kernel modules that only make sense with a specific kernel to that kernel. And similar, using the same string in a unit file or .link file is hence a great way to bind that link file to exactly one kernel version, so that the link or unit file can be shipped within the kernel RPM/DEB, and won't affect any other boots than the ones with the kernel they are shipped with.

So, yes, the kmod thing is really the same thing: it's a way to bind external stuff and a kernel image together, and we use it all over the place, not just for kmods already, but for matching up initrds and whatnot. And with the condition stuff we go one step further, so that we can ship .link files in kernel packages too.

**Comment by poettering** [comment]
I have now force pushed a new version, taking @keszybz patch into account, and adding one on top which adds the version comparison idea. After all after zbigniew's bootspec rework we have a version comparison implementation in the regular codebase, so I figured we can just make use of it now, and implement what has been discussed. PTAL

**Comment by poettering** [comment]
> lgtm except document typos.

hmm, not sure how I forgot to fix those...

Anyway, force pushed a new version now, that fixes this too. PTAL!

**Comment by yuwata** [comment]
Oh, needs to rebase...

**Comment by poettering** [comment]
@yuwata thanks, rebased now. no changes other than that.

**Comment by yuwata** [comment]
Thanks.

**Pull request merged**